test/gdb_script.py

```diff
@@ -0,0 +1,494 @@
+                if error_on_invalid_access:
+                    fail_count += 1
+            except gdb.MemoryError:
+                pass
+
+        # Test valid memory read
+        for addr, value in addr_value_list:
+            val_read = gdb.parse_and_eval("*((int *) 0x%x)" % addr)
+            val_read = int(val_read)
+            assert value == val_read
+
+        # Test invalid memory read
+        for addr in invalid_addr_list:
+            try:
+                gdb.execute("x 0x%x" % addr)
+                if error_on_invalid_access:
+                    fail_count += 1
+            except gdb.MemoryError:
+                pass
+
+        # Test watchpoints
+        access_addr = long(gdb.parse_and_eval("&watchpoint_write_buffer[1]"))
+        bkpt_sizes = [1, 2, 4]
+        bkpt_accesses = ["read", "write", "read_write"]
+        # use "range(-4, 8, 1)" for extended testing
+        bkpt_addresses = [access_addr + offset for offset in range(0, 4, 1)]
+        sizes = [1, 2, 4]
+        accesses = ["read", "write", "read_write"]
+        addresses = [access_addr]
+        generator = product(bkpt_sizes, bkpt_accesses, bkpt_addresses,
+                            sizes, accesses, addresses)
+        for bkpt_size, bkpt_access, bkpt_addr, size, access, addr in generator:
+            gdb_size = size_to_type(bkpt_size)
+            gdb_access = to_gdb_access(bkpt_access)
+            gdb.execute("set var watchpoint_write = %i" %
+                        (1 if has_write(access) else 0))
+            gdb.execute("set var watchpoint_read = %i" %
+                        (1 if has_read(access) else 0))
+            gdb.execute("set var watchpoint_size = %i" % size)
+            gdb.execute("set var write_address = %i" % addr)
+            breakpoint = gdb.Breakpoint("*(%s)0x%x" % (gdb_size, bkpt_addr),
+                                        gdb.BP_WATCHPOINT, gdb_access)
+
+            # Run until breakpoint is hit
+            gdb.execute("c&")
+            event = yield(0.1)
+            bkpt_hit = not is_event_signal(event, "SIGINT")
+
+            # Compare against expected result
+            should_break = should_trigger_break(bkpt_size, bkpt_access,
+                                                bkpt_addr, size, access, addr)
+            valid = valid_watchpoint(bkpt_size, bkpt_access, bkpt_addr)
+
+            if valid and bkpt_hit != should_break:
+                fail_count += 1
+                print("Error - watchpoint problem:")
+                print("  Watchpoint was hit %s" % bkpt_hit)
+                print("  Watchpoint should be hit %s" % should_break)
+                print("  bkpt_size %s, bkpt_access %s, bkpt_address 0x%x, "
+                      "size %s, access %s, addr 0x%x" %
+                      (bkpt_size, bkpt_access, bkpt_addr, size, access, addr))
+                print()
+
+            breakpoint.delete()
+
+        # TODO,future - test reading/writing registers
```

Comment:
TODO,future? Shall we use just TODO ? I would add a name to identify who can provide more context to this todo. Git blame might not always work (moving files, etc).